### PR TITLE
Support ERC20 tokens in searchTx, listenTx and liveTx

### LINKED ISSUE #816: Query transactions for erc20 token
Investigate what information is available from ethereum chain.

We want to allow `searchTx`, `listenTx`, and `liveTx` to detect erc20 token transfers when querying with the `sendFromOrTo` flag.

We should be able to use events/logs to query for this information. For now, consider we have access to a node that persists all logs. In the future, we may have to run our own node for ethereum mainnet/testnet.

TODO: sometime try out on mainnet and decide if this is good enough or we need to rely on etherscan there. For CI and local testnets this should work well without scrapper

### PR DESCRIPTION
**Pull request by webmaster128**
Closes #816

### CODE REVIEW

File: packages/iov-ethereum/src/ethereumcodec.ts [hunk 1]
```diff
@@ -36,6 +36,12 @@ import {
   normalizeHex,
 } from "./utils";
 
+const methodCallPrefix = {
```
Comment (willclarktech) [inline packages/iov-ethereum/src/ethereumcodec.ts]: 👍

File: packages/iov-ethereum/src/ethereumconnection.spec.ts [hunk 2]
```diff
@@ -795,7 +772,124 @@ describe("EthereumConnection", () => {
...
+      });
+      const codec = new EthereumCodec({ erc20Tokens: testConfig.erc20Tokens });
+
+      const profile = new UserProfile();
+      const wallet = profile.addWallet(Secp256k1HdWallet.fromMnemonic(testConfig.mnemonic));
+      const mainIdentity = await profile.createIdentity(wallet.id, testConfig.chainId, HdPaths.ethereum(0));
+
+      const recipient = await randomAddress();
+
+      for (const sendTest of testConfig.erc20TransferTests) {
```
Comment (willclarktech) [inline packages/iov-ethereum/src/ethereumconnection.spec.ts]: I find it odd that we're combining configuration options for a test context with test data in this way.

File: packages/iov-ethereum/src/ethereumconnection.spec.ts [hunk 3]
```diff
@@ -829,10 +925,10 @@ describe("EthereumConnection", () => {
             expect(event.primarySignature.pubkey).toEqual(sender.pubkey);
 
             if (events.length === 3) {
-              // This assumes we get two transactions into one block
-              // A == B < C
-              expect(events[0].height).toEqual(events[1].height);
-              expect(events[2].height).toBeGreaterThan(events[1].height);
+              // The order of the events 0, 1, 2 does not necessarily correspons to the send
```
Comment (willclarktech) [inline packages/iov-ethereum/src/ethereumconnection.spec.ts]: "correspond"

File: packages/iov-ethereum/src/ethereumconnection.spec.ts [hunk 4]
```diff
@@ -871,61 +967,61 @@ describe("EthereumConnection", () => {
...
+        transactionIds.add(postResultB.transactionId);
+        transactionIds.add(postResultC.transactionId);
+
+        // Wait for transaction success
         await postResultA.blockInfo.waitFor(info => !isBlockInfoPending(info));
         await postResultB.blockInfo.waitFor(info => !isBlockInfoPending(info));
-
-        // Post C
-        await connection.postTx(bytesToPostC);
+        await postResultC.blockInfo.waitFor(info => !isBlockInfoPending(info));
```
Comment (willclarktech) [inline packages/iov-ethereum/src/ethereumconnection.spec.ts]: Suggest using a `Promise.all` here.

File: packages/iov-ethereum/src/ethereumconnection.spec.ts [hunk 5]
```diff
@@ -1017,19 +1113,244 @@ describe("EthereumConnection", () => {
...
+        const sendC: SendTransaction = {
+          kind: "bcp/send",
+          creator: sender,
+          recipient: recipientAddress,
+          fee: {
+            gasPrice: testConfig.gasPrice,
+            gasLimit: testConfig.gasLimit,
+          },
+          ...testConfig.erc20TransferTests[0],
+        };
```
Comment (willclarktech) [inline packages/iov-ethereum/src/ethereumconnection.spec.ts]: Does this really need to be defined three times?

File: packages/iov-ethereum/src/ethereumconnection.spec.ts [hunk 6]
```diff
@@ -1017,19 +1113,244 @@ describe("EthereumConnection", () => {
...
+            expect(event.transaction.creator).toEqual(sender);
+            expect(event.primarySignature.pubkey).toEqual(sender.pubkey);
+
+            if (events.length === 3) {
+              const receivedIds = new Set(events.map(e => e.transactionId));
+              expect(receivedIds).toEqual(transactionIds);
+
+              subscription.unsubscribe();
+              connection.disconnect();
+              done();
```
Comment (willclarktech) [inline packages/iov-ethereum/src/ethereumconnection.spec.ts]: Did you consider doing this Promise-style?

File: packages/iov-ethereum/src/ethereumconnection.ts [hunk 7]
```diff
@@ -702,23 +826,106 @@ export class EthereumConnection implements BcpConnection {
...
+      }
     }
+
+    return out;
+  }
+
+  private async searchSendTransactionsByAddressInLogs(
+    address: Address,
+    minHeight: number,
+    maxHeight: number,
```
Comment (willclarktech) [inline packages/iov-ethereum/src/ethereumconnection.ts]: Could these have default values?

File: packages/iov-ethereum/src/ethereumconnection.ts [hunk 8]
```diff
@@ -702,23 +826,106 @@ export class EthereumConnection implements BcpConnection {
...
+    // console.log(erc20TransferLogsResponse.result);
+    // console.log(
+    //   erc20TransferLogsResponse.result.map((row: any) => {
+    //     return {
+    //       height: decodeHexQuantity(row.blockNumber),
+    //       contractAddress: row.address,
+    //       topics: row.topics,
+    //     };
+    //   }),
+    // );
```
Comment (willclarktech) [inline packages/iov-ethereum/src/ethereumconnection.ts]: `console.log`s

File: packages/iov-ethereum/src/ethereumconnection.ts [hunk 9]
```diff
@@ -702,23 +826,106 @@ export class EthereumConnection implements BcpConnection {
...
+        },
+      ],
+      id: 7,
+    });
+    if (isJsonRpcErrorResponse(erc20TransferLogsResponse)) {
+      throw new Error(JSON.stringify(erc20TransferLogsResponse.error));
+    }
+
+    if (!Array.isArray(erc20TransferLogsResponse.result)) {
+      throw new Error("Got unuepected type of result");
```
Comment (willclarktech) [inline packages/iov-ethereum/src/ethereumconnection.ts]: "Expected result to be an array"?

File: packages/iov-ethereum/src/ethereumconnection.ts [hunk 10]
```diff
@@ -702,23 +826,106 @@ export class EthereumConnection implements BcpConnection {
...
+    //     return {
+    //       height: decodeHexQuantity(row.blockNumber),
+    //       contractAddress: row.address,
+    //       topics: row.topics,
+    //     };
+    //   }),
+    // );
+
+    const ids = erc20TransferLogsResponse.result.map(row => Parse.transactionId(row.transactionHash));
+    // query all in paralel
```
Comment (willclarktech) [inline packages/iov-ethereum/src/ethereumconnection.ts]: "parallel"

File: packages/iov-ethereum/src/ethereumconnection.spec.ts (hunk 2 above)
Comment (webmaster128) [reply packages/iov-ethereum/src/ethereumconnection.spec.ts]: Me too. But I do not have a great idea how to change that. I mean we can enforce a list length of 2 and can have a test for the primary and a test for the secondary token. Not sure how that scales.

File: packages/iov-ethereum/src/ethereumconnection.spec.ts (hunk 5 above)
Comment (webmaster128) [reply packages/iov-ethereum/src/ethereumconnection.spec.ts]: All 3 are indeed equal in this test. Given that there is no nice native object cloning, I prefer to roll out this code.

File: packages/iov-ethereum/src/ethereumconnection.spec.ts (hunk 6 above)
Comment (webmaster128) [reply packages/iov-ethereum/src/ethereumconnection.spec.ts]: Wrapping those event-based tests into promises hides a lot of the underlying detail and removes control, which is important for debugging. Thus all the liveTx/listenTx tests are written this way.

File: packages/iov-ethereum/src/ethereumconnection.ts (hunk 7 above)
Comment (webmaster128) [reply packages/iov-ethereum/src/ethereumconnection.ts]: Why would we need them? All callers set the values explicitly, which makes this function simpler.

File: packages/iov-ethereum/src/ethereumconnection.ts (hunk 9 above)
Comment (webmaster128) [reply packages/iov-ethereum/src/ethereumconnection.ts]: Thanks for suggesting an alternative!

File: packages/iov-ethereum/src/ethereumconnection.spec.ts (hunk 2 above)
Comment (webmaster128) [reply packages/iov-ethereum/src/ethereumconnection.spec.ts]: But what I find worse the current version is creating tests dynamically (call `describe` or `it` based on runtime data). We do this in iov-tendermint-rpc and it is very hard to follow the test code.

File: packages/iov-ethereum/src/ethereumconnection.spec.ts (hunk 5 above)
Comment (willclarktech) [reply packages/iov-ethereum/src/ethereumconnection.spec.ts]: I was assuming they were going to be consumed by a pure function, but maybe that's not a safe assumption to make.
